### ISSUE TITLE
SQL injection possible with limit() on MySQL

### ISSUE THREAD

**Issue by mpetrovich** [propelorm/Propel2#1463]
**The `limit()` query method is susceptible to catastrophic SQL injection with MySQL.**

For example, given a model `User` for a table `users`:
```php
UserQuery::create()->limit('1;DROP TABLE users')->find();
```
This will drop the `users` table!

The cause appears to be a lack of integer casting of the limit input in either [`Propel\Runtime\ActiveQuery\Criteria::setLimit()`](https://github.com/propelorm/Propel2/blob/master/src/Propel/Runtime/ActiveQuery/Criteria.php#L1260-L1266) or in [`Propel\Runtime\Adapter\Pdo\MysqlAdapter::applyLimit()`](https://github.com/propelorm/Propel2/blob/master/src/Propel/Runtime/Adapter/Pdo/MysqlAdapter.php#L122-L129). The code comments there seem to imply that casting was avoided due to overflow issues with 32-bit integers.

This is surprising behavior since one of the primary purposes of an ORM is to prevent basic SQL injection.

**This affects all versions of Propel: 1.x, 2.x, and 3.**

**Comment by marcj** [comment]
Care to provide a PR?

**Comment by mpetrovich** [comment]
Happy to! Coming soon.

**Comment by halfer** [comment]
@mpetrovich: good spot. However, food for thought - it is usual practice to notify a project of security problems privately, rather than making it public in the first instance.

**Comment by mpetrovich** [comment]
Good point, pardon my rush to post it here. Please delete these issues in the meantime if you’d like—I’m working on a pull request to address this issue.

**Comment by mpetrovich** [comment]
PR with a fix: https://github.com/propelorm/Propel2/pull/1464

**Comment by halfer** [comment]
@marcj: I think this should this get an alert for the relevant versions [on the Symfony security checker](https://security.sensiolabs.org/). Do you agree that all prior versions were vulnerable? I suspect the PR to that service needs to specify what versions are affected, and thus it needs to be accurate.

Does this need a CVE reference? I don't know how to go about getting that assigned.

**Comment by halfer** [comment]
@mpetrovich:

Partly to answer my own question, I see you have a bug report for each of the major versions. Thanks.

I have prepared a [suggested change](https://github.com/FriendsOfPHP/security-advisories/compare/master...halfer:master) and will submit that as a PR to the sec database tomorrow (Sunday 18th Feb). If you are around and are able to double-check the contents, that'd be great.

In particular, I don't know if `>=2` is considered in Composer terms to include all the alpha releases in this project. Unfortunately I cannot test this until the changes are merged to the FriendsOfSymfony repo. Comments on this welcome.

Presently for Propel1 I have marked everything including the current release as vulnerable. If a new release is made, ping me and I will push a new change.

**Comment by mpetrovich** [comment]
@halfer:

> In particular, I don't know if >=2 is considered in Composer terms to include all the alpha releases in this project.

I believe so. According to [the Composer docs](https://getcomposer.org/doc/articles/versions.md#tilde-version-range-), any alpha version like `2.x-alpha-y` is strictly before `2.0`, but both are still in the `2.x` range and should be satisfied by any rule like `~2` or `>=2`.

> Presently for Propel1 I have marked everything including the current release as vulnerable. If a new release is made, ping me and I will push a new change.

I've made pull requests [1] [2] to fix the same vulnerabilities in Propel v1 and v3, so you should be able to restrict the vulnerability version range to everything up to the current releases.
[1] https://github.com/propelorm/Propel/pull/1054
[2] https://github.com/propelorm/Propel3/pull/74

**Comment by halfer** [comment]
Marvellous, thanks @mpetrovich. I've set the Propel1 constraints to `<=1.7.1`, so if a new release is made, I am assuming it will contain your fix.

I'll raise a PR just now and will liaise with maintainers to get this merged into the advisories db. Once that is done I'll build a demo Propel1 and Propel2 project to make sure the warnings come up on the latest releases.
